packages/opentelemetry-resource-detector-aws/src/detectors/AwsLambdaDetector.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Detector,
+  Resource,
+  CLOUD_RESOURCE,
+  ResourceDetectionConfig,
+} from '@opentelemetry/resources';
+
+/**
+ * The AwsLambdaDetector can be used to detect if a process is running in AWS Lambda
+ * and return a {@link Resource} populated with data about the environment.
+ * Returns an empty Resource if detection fails.
+ */
+export class AwsLambdaDetector implements Detector {
+  async detect(_config?: ResourceDetectionConfig): Promise<Resource> {
```

Comment:
Hmm - I realized this approach doesn't work so well because in lambda we'd generally advice to use a synchronous wrapper for initializing OTel, like https://github.com/open-telemetry/opentelemetry-lambda/blob/main/nodejs/packages/layer/src/wrapper.ts

But can't await on a resource here. If I could `registerInstrumentations` first to at least get code patched before an await, it'd be ok but otherwise I think code can get loaded before patch.

Any advice on a good approach? Should I just extract this into `export function detectLambdaResourceSync()`?

Reply:
You can register instrumentations before registering the tracer factory. The instrumentations will use the global tracer factory by default which is a proxy tracer factory which returns proxy tracers which proxy calls to a noop tracer. When a tracerfactory is registered, the tracers already acquired become "active" tracers instead of noop tracers.

Reply:
the real solution for this problem is to make it so that resources can be appended to or similar. I know there is some disucussion around this in the spec. then you can load your sync resources at startup, then have the async ones show up when they are ready

Reply:
Thanks @dyladan - I came up with this pattern calling `registerInstrumentations` twice

https://github.com/open-telemetry/opentelemetry-lambda/pull/30/files#diff-fc60c587485bcf624645ecb4ab13cc4393085d3f28ff0490eb275203abdf4d16R34

I first tried without the second `registerInstrumentations` to see if `tracerProvider.register()` is enough to make the tracers become "active", but it didn't seem to work. But this double-registration pattern doesn't seem too bad to me either.

If that looks ok, I think this PR is ready for review!

Reply:
That's strange and sounds like a bug somewhere. It should be really enough to call `registerInstrumentations()` once without passing a `TracerProvider`.

Reply:
`registerInstrumentations` is needed, but the tracerProvider is optional and should be user either the one you passed or it will use the global one.
see here
https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-instrumentation/src/autoLoader.ts#L38
so just let be clear, did you try to call it `registerInstrumentations` without provider and it didn't work ?

With regards to order `registerInstrumentations(...)` and `provider.register(...)` it shouldn't matter, both should work fine. If it doesn't work then something might be wrong with proxy or api global (just guessing)

Reply:
@obecny The order I did is

1) `registerInstrumentations` without `tracerProvider`
2) `tracerProvider.register()`

And it didn't seem to work yet. So I added 

3) `registerInstrumentations` with `tracerProvider`

I didn't try 2) `registerInstrumentations` without `tracerProvider` but don't really need to since I have the provider ready when doing that :)

Reply:
in simple setups were it is ensured that version of all components matches and there is a single place for all instrumentations to be registered,.. it's mostly don't care how this is done.
But in a more complex setup where e.g some 3rd party components are be involved which may want to install some more instrumentations,... use OTel directly,.. passing around a TracerProvider instances may easily result in problems.
Some locations use the global one, others use that one passed around.

Reply:
Yeah - I can try a second register without passing a provider and see how it goes. Either way I call provider.register so I guess it'll be mostly the same.

Anyways these are my wrapper in the Lambda repo, but appreciate a review of this independent code :)